synclog to SQL

sravfeyn opened the pull request:
@snopoke This is a preliminary attempt to migrate synclogs to SQL. This is work in progress and there is still good amount of work left out. Does this look like a correct direction?

sravfeyn committed:
WIP; synclog to SQL

sravfeyn committed:
self

emord commented:
Looks like this is only looking at SQL right now. I think the original intention is to try to find the synclog in SQL then if it's not there, try to find it in couch, so that we could cleanly move from couch to sql without a bunch of 412s. Any reason for going in this direction?

sravfeyn committed:
few fixes

sravfeyn committed:
commit migration file

sravfeyn committed:
handle unknown doc id better

sravfeyn committed:
handle bad synclog token correctly

sravfeyn committed:
date nullable

sravfeyn committed:
fix tests

sravfeyn committed:
domain nullable, but not user_id

sravfeyn committed:
WIP; migration script

sravfeyn commented:
@emord Yeah, I will extend querying so that it looks in couch if a doc is not found in SQL. I am also writing a couch to SQL migration script.

emord commented:
>  I am also writing a couch to SQL migration script.

I don't think we need to do that. Synclogs are deleted after 60 days (I think 30 days for ICDS) so if there's a fallback to couch we can just delete the databases in a couple of months once everything is being read from SQL.

sravfeyn committed:
upsert

sravfeyn committed:
use hex value

sravfeyn committed:
couch to SQL

sravfeyn committed:
remove synclog couch dump test

sravfeyn committed:
Merge branch 'master' into sr-pg-synclogs

sravfeyn committed:
allow lazy migration

sravfeyn committed:
deal with todos

sravfeyn commented on corehq/apps/dump_reload/tests/test_couch_dump_load.py:
```diff
@@ -193,40 +193,6 @@ def test_web_user(self):
 
         self._dump_and_load([web_user], [other_user])
 
-    def test_sync_log(self):
```
Should I move couch synclog datadump to SQL, is it useful?

sravfeyn commented:
@snopoke I am mostly done with what I planned to do, there are couple of comments I have added where your inputs are useful. 

I have fixed most of tests, I think there will be few more (I am waiting for CI). If everything is good, we can try this on staging next.

Are there any things in migration this PR is missing?

sravfeyn commented:
> I don't think we need to do that. Synclogs are deleted after 60 days (I think 30 days for ICDS) so if there's a fallback to couch we can just delete the databases in a couple of months once everything is being read from SQL.
@emord, ha, okay! I already wrote the script in case we want to do it more proactively.

snopoke commented on corehq/ex-submodules/casexml/apps/phone/exceptions.py:
```diff
@@ -39,6 +39,10 @@ class InvalidSyncLogException(Exception):
     pass
 
 
+class MissingSyncLogSQLException(Exception):
```
Is it necessary to have a different exception? Can't we just use the `MissingSyncLog` exception?

sravfeyn replied:
Yeah, I think we could, I will check

snopoke commented on corehq/ex-submodules/casexml/apps/phone/management/commands/invalidate_sync_heads.py:
```diff
@@ -15,17 +17,12 @@ def add_arguments(self, parser):
+        for synclog in synclogs:
+            doc = properly_wrap_sync_log(synclog.doc)
+            doc.case_ids_on_phone = {'broken to force 412'}
+        bulk_update_helper(synclogs)
```
This should continue to support couch until we're fully migrated.

sravfeyn replied:
Cool, will do

snopoke commented on corehq/ex-submodules/casexml/apps/phone/models.py:
```diff
@@ -354,6 +355,53 @@ def test_only_get_dependent_cases_on_phone(self):
+    date = models.DateTimeField(db_index=True, null=True, blank=True)
+    synclog_id = models.UUIDField(unique=True, primary_key=True)
+    # needs to be a foreign key?
+    previous_synclog_id = models.CharField(max_length=255, default=None, null=True, blank=True)
```
Could be but probably not necessary. Any reason it's not a UUID field?

snopoke commented on corehq/ex-submodules/casexml/apps/phone/models.py:
```diff
@@ -354,6 +355,53 @@ def test_only_get_dependent_cases_on_phone(self):
+    domain = models.CharField(max_length=255, null=True, blank=True, default=None, db_index=True)
+    user_id = models.CharField(max_length=255, default=None, db_index=True)
+    date = models.DateTimeField(db_index=True, null=True, blank=True)
+    synclog_id = models.UUIDField(unique=True, primary_key=True)
```
If this is the primary key it should be the first field (by convention)

sravfeyn replied:
Is it possible to make this something other than UUID right now because this comes from couch during the couch to sql migration. After the migration is over it might be easier to replace this with another incremental UUID  (the only refs that need update is previous_synclog_id)

Re. default, the UUID manually when instantiating an object otherwise there would be an exception, so is a default still necessary?

emord replied:
> Is it possible to make this something other than UUID right now because this comes from couch during the couch to sql migration

We don't need to migrate this. All read queries should do something like:

```python
def get_synclog(synclog_id):
    try:
        return SyncLogSQL.get(synclog_id) 
    except:
        return CouchSyncLog.get(synclog_id)
```

That way we can fallback to the old synclogs without an active migration.

All write queries should only write to SQL.

> Re. default, the UUID is manually set when instantiating an object otherwise there would be an exception, so is a default still necessary?

I don't see where the UUID is being set manually. Can you link me to that? I don't think there's any harm with setting a default

sravfeyn replied:
Sorry, I could have been more clear.

I already implemented the lazy migration (querying from couch and SQL for read but not for write instead of onetime migration). My above response was to your suggestion of making synclog_id (UUID) a sequential ID instead of couch like ID for better query performance. 

The manual UUID setting is done [here](https://github.com/dimagi/commcare-hq/pull/19391/files#diff-a7f81fae5762e257cd3241a8f458496dR400). Also if you look at that function it relies on old couch synclog IDs which are not sequential. So it's tricky to replace them with new sequential IDs as part of the lazy migration since it would be difficult to keep track of correspondence of couch and SQL synclogs. 

Once the lazy migration is fully complete (after 60 days), we can add a new ID column that is sequential and do a one time migration to replace the current synclog_id (UUID that is not sequential, that came from couch), it should be possible to do this very efficiently since all the all calculations will partain to the synclog sql table itself.

emord replied:
Couch IDs do use a sequential uuids http://docs.couchdb.org/en/2.1.1/config/misc.html?highlight=sequential#uuids/algorithm

Anyways I think that may end up being a later optimization, but we may want to default to uuid1 (instead of uuid4) since it will be less random.

>  Also if you look at that function it relies on old couch synclog IDs which are not sequential

From my understanding, SyncLogs are never updated after it's sent to the user. So all new writes are only going to SQL and we won't write any docs that are in couch to postgres. When a user syncs after this PR, it will look like:

1) User "syncs to server"
2) In user's request there is a synclog_id ABC.
3) This code looks in SQL for synclog ABC, but it's not there. Looks in Couch gets synclog
4) restore code figures out what cases/UCRs to send back to the user
5) a new synclog DEF is generated with what information is being sent to the user.
6) synclog DEF is saved to SQL
7) restore is sent to user.
8) On next sync, synclog DEF is sent from user to HQ and all subsequent syncs are only relying on SQL

Synclog ABC will never be stored in SQL so there shouldn't be any reason that we're restricted due to couch ids

sravfeyn replied:
Ah, that's right. Switching to uuid1, I had the migration script playing in my head when I thought about this.

snopoke commented on corehq/ex-submodules/casexml/apps/phone/tests/test_ota_restore.py:
```diff
@@ -105,19 +107,17 @@ def tearDown(self):
 @override_settings(CASEXML_FORCE_DOMAIN_CHECK=False)
 class OtaRestoreTest(BaseOtaRestoreTest):
 
+    def _get_the_only_synclog(self):
```
Since this is called `get_the_ONLY_synclog` should it raise an exception if there are more than one? Otherwise maybe it should be called `get_the_first_synclog`?

sravfeyn replied:
These tests are written such a way that there would be only one synclog, and the query was relying on that fact, so exception is necessary I will change the name.

snopoke commented on corehq/ex-submodules/casexml/apps/phone/utils.py:
```diff
@@ -21,6 +21,7 @@
 
 
 def delete_sync_logs(before_date, limit=1000, num_tries=10):
+    # Todo: convert to SQL including get_synclog_ids_before_date
```
We don't need to fully delete them but we should blank out the `doc` field to reclaim some space.

snopoke replied:
One benefit of not deleting them is that the synclog report would work again. Perhaps we can compromise and clear the json data after 30 days (or on the next sync) but keep the metadata for a longer period.

emord replied:
Yeah  that makes sense. Though will that be fixed in the data warehouse? I recall seeing something about a synclogfact table.

sravfeyn replied:
I think we can let this do couch deletion (instead of SQL) till we have around 30 days of SQL synclogs, so that this will continue to be useful for deleting couch docs. I will convert this to SQL version in a separate PR. Does that sound right?

snopoke replied:
Seems OK I guess.

snopoke commented on corehq/pillows/synclog.py:
```diff
@@ -23,6 +23,7 @@ def get_user_sync_history_pillow(pillow_id='UpdateUserSyncHistoryPillow', **kwar
     """
     This gets a pillow which iterates through all synclogs
     """
+    #Todo; convert to SQL?
```
yes

sravfeyn replied:
Cool. Is there an example of SQL pillow processor?

snopoke replied:
the 'processor' won't change much. The big difference is that the pillow will need to use a `KafkaChangeFeed` instead of a `CouchChangeFeed`. Here's an example: https://github.com/dimagi/commcare-hq/blob/3e88dbe4554316981e7f9a43fde34aa8870702c9/corehq/pillows/app_submission_tracker.py#L24

snopoke replied:
@sravfeyn is this going to be included in this PR?

snopoke commented on corehq/pillows/synclog.py:
```diff
@@ -41,7 +42,7 @@ def get_user_sync_history_pillow(pillow_id='UpdateUserSyncHistoryPillow', **kwar
 
     def process_change(self, pillow_instance, change):
-
+        #Todo; convert to SQL?
```
yes

snopoke replied:
also still outstanding

snopoke commented on corehq/ex-submodules/casexml/apps/phone/models.py:
```diff
@@ -354,6 +355,53 @@ def test_only_get_dependent_cases_on_phone(self):
+            (format, format)
+            for format in [LOG_FORMAT_LEGACY, LOG_FORMAT_SIMPLIFIED, LOG_FORMAT_LIVEQUERY]
+        ]
+    )
```
There are some other fields on SyncLogs that could be useful to keep:
* build_id
* duration
* last_submitted
* had_state_error
* error_date
* error_hash

snopoke reviewed (changes requested):
My biggest question here is whether or not we want to make this a sharded model. My opinion is that we should. I would also suggest that we do the sharding on `user_id` rather than `synclog_id`.

emord reviewed (commented):
> My biggest question here is whether or not we want to make this a sharded model. My opinion is that we should. 

Either sharding or partitioning seems somewhat necessary to me.

Sharding - Can distribute it to multiple servers, but it competing for same resources as cases/forms/ledgers.
Partitioning - Could be in it's own database. Unable to distribute over multiple servers. (feels a bit easier to me as with architect it's somewhat transparent, but maybe just because I haven't really added any sharded models)

Personally, I'm somewhat skeptical we'd need performance of sharding on a database that will likely only have 7 million rows at max assuming 5 synclogs per user, which seems unlikely since we also delete old synclogs on a new sync. Though I've certainly been wrong about this sort of thing before.

> I would also suggest that we do the sharding on user_id rather than synclog_id.

Curious why you would pick user_id instead of synclog_id. The only instance of searching by user_id in real code was for the update user sync history pillow reindexer, which is only run if we need to change an ES index I think. And may get rewritten after teh data warehouse changes the format for synclog history reports (not sure about that though?)

I think that the majority of synclogs are retrieved in restore by only synclog_id so if we partition by that the request would only go to one database instead of all 10. If we do end up partitioning by user_id we should make sure we can feed user_id to all queries.

snopoke commented:
> > My biggest question here is whether or not we want to make this a sharded model. My opinion is that we should.
> 
> Either sharding or partitioning seems somewhat necessary to me.
> 
> Sharding - Can distribute it to multiple servers, but it competing for same resources as cases/forms/ledgers.
> Partitioning - Could be in it's own database. Unable to distribute over multiple servers. (feels a bit easier to me as with architect it's somewhat transparent, but maybe just because I haven't really added any sharded models)
> 
> Personally, I'm somewhat skeptical we'd need performance of sharding on a database that will likely only have 7 million rows at max assuming 5 synclogs per user, which seems unlikely since we also delete old synclogs on a new sync. Though I've certainly been wrong about this sort of thing before.

You're probably right, the sync rate isn't that high that we'd really need to shard so partitioning by date is probably fine and as you say it makes it easier to have it in a separate DB.

> > I would also suggest that we do the sharding on user_id rather than synclog_id.
> 
> Curious why you would pick user_id instead of synclog_id. The only instance of searching by user_id in real code was for the update user sync history pillow reindexer, which is only run if we need to change an ES index I think. And may get rewritten after teh data warehouse changes the format for synclog history reports (not sure about that though?)
> 
> I think that the majority of synclogs are retrieved in restore by only synclog_id so if we partition by that the request would only go to one database instead of all 10. If we do end up partitioning by user_id we should make sure we can feed user_id to all queries.

I don't think there's ever a situation where we don't know which user we're looking for as well. During a sync / form submission we already know which user it is based on the auth.

The main reason I suggested it is for the report since it would allow us to query a single DB instead of multiple.

All this is moot if we're not going to shard.

emord commented:
> You're probably right, the sync rate isn't that high that we'd really need to shard so partitioning by date is probably fine and as you say it makes it easier to have it in a separate DB.

We could do that, but if we do, I think we would also want to have a date filter so that we only look at a subset of filters.

I was thinking about partitioning based on string_firstchars http://architect.readthedocs.io/features/partition/postgresql.html#string-firstchars but now I realize that may only work for strings, might be good to test. Could also do it based on user_id which is a charfield. Maybe lastchars is actually more distributed (since couch uses a sequential uuid algorithm)?

We could also add another field for a random small integer that says what partition it goes to and shard based on range of the integer. For example if we had a 16 bit integer and chose a range of 1024 it would split it up in 32 databases, but we'd have to pass this in the restore. Likely not feasible.

> The main reason I suggested it is for the report since it would allow us to query a single DB instead of multiple.

> All this is moot if we're not going to shard.

As noted above we can also partition based on user_id (assuming that first or last chars for the id are distributed evenly) or a random integer

sravfeyn committed:
typo

sravfeyn committed:
reuse exceptiption

sravfeyn committed:
note

sravfeyn committed:
readd couch

sravfeyn committed:
rename

sravfeyn committed:
additional fields

sravfeyn committed:
wrap

sravfeyn committed:
query tests

snopoke commented:
synclog also get updated during form submission (adding new cases etc).

sravfeyn committed:
fix mistakes

sravfeyn committed:
uuid1

sravfeyn committed:
import

sravfeyn committed:
WIP; enable using separte synclogs SQL DB

sravfeyn committed:
use partionedmodel

sravfeyn committed:
no need for partionedmodel

sravfeyn committed:
default setting

sravfeyn committed:
remove import

sravfeyn committed:
Merge pull request #19562 from dimagi/sr-router

WIP; enable using separte synclogs SQL DB

sravfeyn committed:
commit migration

sravfeyn commented on corehq/ex-submodules/casexml/apps/phone/dbaccessors/sync_logs_by_user.py:
```diff
@@ -22,7 +30,20 @@ def get_last_synclog_for_user(user_id):
 
-def get_synclogs_for_user(user_id, limit=10):
+def get_synclogs_for_user(user_id, limit=10, wrap=True):
+    synclogs = SyncLogSQL.objects.filter(user_id=user_id).order_by('date')[:limit]
```
Is that necessary, if we are going to prune them at DB level?

sravfeyn replied:
@snopoke  I think this is the only comment pending

snopoke replied:
This is OK for now

sravfeyn committed:
setting validation

sravfeyn committed:
merge master

sravfeyn committed:
remove extra valdiation

emord commented on corehq/ex-submodules/casexml/apps/phone/models.py:
```diff
@@ -354,6 +356,94 @@ def test_only_get_dependent_cases_on_phone(self):
+    return synclog
+
+
+@architect.install('partition', type='range', subtype='date', constraint='day', column='date')
```
If we're keeping these around indefinitely I think we should change this constraint to either week or month. 

I just ran into an issue of poor insert performance while testing one of the new UCRs and will be changing that to month. Good rule of thumb is ~100 partitions according to [docs](https://www.postgresql.org/docs/10/static/ddl-partitioning.html)

month gives us ~8 years of data before it should impact performance, weeks gives ~2 years. If it's ever ok to prune this to < 3 months, then day should be good

sravfeyn replied:
Hey, so per Simon's below comment we are gonna keep this to day, right?

emord replied:
@sravfeyn I think he meant to change it to week

snopoke replied:
yes, change to week

sravfeyn replied:
Updated

snopoke commented:
I think it's fine to prune it to < 3 months but weeks still seems fine for
that (12 tables instead of 90).

sravfeyn committed:
fix order

sravfeyn committed:
merge master

sravfeyn committed:
change to week

snopoke commented on corehq/sql_db/__init__.py:
```diff
@@ -0,0 +1,38 @@
+
+    def ready(self):
+        super(SQLDBAppConfig, self).ready()
+        register()(custom_db_checks)
```
Is this line necessary? The function already has the `register` decorator and since this is in `__init__` it will get processed during import.

sravfeyn replied:
Thanks, removed it!

snopoke commented on corehq/sql_db/routers.py:
```diff
@@ -45,6 +46,8 @@ def allow_migrate(db, app_label):
         db_alias = get_icds_ucr_db_alias()
         return db_alias and db_alias == db
+    elif app_label == SYNCLOGS_APP:
+        return hasattr(settings, "SYNCLOGS_SQL_DB_ALIAS") and db == settings.SYNCLOGS_SQL_DB_ALIAS
```
`settings` will always have `SYNCLOGS_SQL_DB_ALIAS` so the `hasattr` check isn't necessary

sravfeyn replied:
Thanks, fixed.

snopoke commented on corehq/ex-submodules/casexml/apps/phone/tests/dbaccessors.py:
```diff
@@ -1,13 +1,9 @@
-        reduce=False,
-    )]
-    return iter_docs(SyncLog.get_db(), all_sync_log_ids)
+	for synclog in SyncLogSQL.objects.all():
```
@sravfeyn needs fixing

snopoke commented on corehq/warehouse/dbaccessors.py:
```diff
@@ -68,23 +68,16 @@ def get_synclog_ids_by_date(start_datetime, end_datetime):
         # Skip the record if the datetime is equal to the start because this should return
         # records with an exclusive start date.
-        if result_modified_datetime == json_start_datetime:
+        if result_modified_datetime == start_datetime:
```
not necessary since we're using `gt` filter

snopoke commented on corehq/warehouse/dbaccessors.py:
```diff
@@ -68,23 +68,16 @@ def get_synclog_ids_by_date(start_datetime, end_datetime):
-    )
-    for result in results:
-        result_modified_datetime = result['key'][0]
+    synclogs = SyncLogSQL.objects.filter(date__gt=start_datetime, date__lt=end_datetime)
```
could do a `values_list` query to only return the ID

snopoke commented:
@sravfeyn I think this is also missing the migration to create the table partitions

sravfeyn committed:
remove unnecessary call

sravfeyn committed:
remove hasattr check

sravfeyn committed:
covert 8 spaces to 4

sravfeyn committed:
skip check

sravfeyn committed:
simpler query

sravfeyn committed:
basic test synclog pillow; fail

sravfeyn committed:
covert pillow to kafka SQL

sravfeyn committed:
flat

sravfeyn committed:
remove limit

sravfeyn committed:
correction

emord commented on corehq/ex-submodules/casexml/apps/phone/dbaccessors/sync_logs_by_user.py:
```diff
@@ -22,7 +30,20 @@ def get_last_synclog_for_user(user_id):
-def get_synclogs_for_user(user_id, limit=10):
+def get_synclogs_for_user(user_id, limit=10, wrap=True):
+    synclogs = SyncLogSQL.objects.filter(user_id=user_id).order_by('date')[:limit]
+    if synclogs:
```
If a user has only 1 synclog in SQL and 9 in couch then this will only return the 1 SQL synclog instead of the 10 defined in limit. I think that's probably ok for now

sravfeyn committed:
correction

sravfeyn committed:
import

sravfeyn committed:
kafka

sravfeyn committed:
clean up test

sravfeyn committed:
correct lookup

sravfeyn committed:
more kafka corrections

sravfeyn committed:
admin model

sravfeyn committed:
update asserts

sravfeyn committed:
1

sravfeyn commented:
@snopoke I have converted the pillow to kafka/sql. Does this look ready to be deployed on staging?

snopoke commented on corehq/ex-submodules/casexml/apps/phone/migrations/0002_synclogsql.py:
```diff
@@ -0,0 +1,37 @@
+# Generated by Django 1.11.8 on 2018-02-09 12:14
+from __future__ import unicode_literals
+
+import uuid
```
@sravfeyn you should apply this change

snopoke replied:
@sravfeyn you should apply this change

snopoke commented on corehq/ex-submodules/casexml/apps/phone/tests/test_synclog_pillow.py:
```diff
@@ -0,0 +1,56 @@
+        synclog = SyncLog(domain=self.domain.name, user_id=self.ccuser._id,
+                          date=datetime.datetime(2015, 7, 1, 0, 0)),
+        synclog.save()
+        message = consumer.next()
```
@sravfeyn you should use `next(consumer)`

snopoke commented on corehq/ex-submodules/casexml/apps/phone/tests/test_synclog_pillow.py:
```diff
@@ -0,0 +1,56 @@
+        self.assertEqual(change_meta.document_id, synclog.synclog_id)
+        self.assertEqual(change_meta.domain, self.domain.name, )
+
+        self.assertEqual(self.ccuser.reporting_metadata.last_syncs, [])
```
shouldn't this have something in it now? (after you re-load the user)

snopoke commented on corehq/ex-submodules/casexml/apps/phone/admin.py:
```diff
@@ -24,4 +24,11 @@ class OwnershipCleanlinessFlagAdmin(admin.ModelAdmin):
 
+class SyncLogSQLAdmin(admin.ModelAdmin):
+    model = SyncLogSQL
+    list_display = ['synclog_id', 'domain', 'user_id', 'date', 'doc']
```
`doc` is huge so probably shouldn't be displayed in the list

snopoke reviewed (commented):
Looks good, a few small comments but nothing blocking

sravfeyn committed:
few changes

snopoke commented on corehq/pillows/synclog.py:
```diff
@@ -30,7 +30,7 @@ def get_user_sync_history_pillow(
         topics=[topics.SYNCLOG_SQL], group_id=SYNCLOG_SQL_USER_SYNC_GROUP_ID,
         num_processes=num_processes, process_num=process_num)
-    checkpoint = KafkaPillowCheckpoint('synclog-user-sync', [topics.SYNCLOG_SQL])
+    checkpoint = KafkaPillowCheckpoint('synclog', [topics.SYNCLOG_SQL])
```
why change this? In most cases we make this equal to `pillow_id`

snopoke replied:
I see you were just copying from what was there before. I think better to change this to be `pillow_id`

sravfeyn replied:
Reverted and fixed the test data instead

sravfeyn committed:
change checkpoint_id

sravfeyn committed:
user pillow_id

snopoke commented on testapps/test_pillowtop/tests/data/all-pillow-meta.json:
```diff
@@ -219,7 +219,7 @@
         "advertised_name": "UpdateUserSyncHistoryPillow",
         "change_feed_type": "KafkaChangeFeed",
-        "checkpoint_id": "synclog-user-sync",
+        "checkpoint_id": "update-user-sync-history",
```
confused why this isn't `UpdateUserSyncHistoryPillow`

sravfeyn replied:
I was trying to reverse guess the logic by looking at other pillow names in this file. I assumed there might be a function that coverts class names like `UpdateUserSyncHistoryPillow` to `update-user-sync-history`. But it doesn't look like so. I have updated

sravfeyn committed:
correct id

sravfeyn commented:
DB creation checklist after which this can be merged on master

- [ ] enikshay
- [ ] icds-new
- [ ] motech
- [ ] production
- [ ] staging
- [ ] icds
- [ ] l10k
- [ ] pna
- [ ] softlayer
- [ ] swiss

sravfeyn commented:
@snopoke  Hi Simon, I have created synclogs DB on all envs and deployed localsettings to include synclogs DB. I have also tested on staging that the sync works, that the log gets saved in SQL and that user sync history metadata gets updated (via Kafka). 

So this is good to go. Once merged, I will deploy this on ICDS.

snopoke committed:
Merge branch 'master' into sr-pg-synclogs

snopoke commented on corehq/ex-submodules/casexml/apps/phone/management/commands/migrate_synclogs_to_sql.py:
```diff
@@ -0,0 +1,91 @@
+from dimagi.utils.couch.database import iter_docs_with_retry
+
+
+class Command(BaseCommand):
```
@sravfeyn are you planning on using this? If not you might as well remove it from the code.

snopoke commented on corehq/ex-submodules/casexml/apps/phone/models.py:
```diff
@@ -354,6 +358,105 @@ def test_only_get_dependent_cases_on_phone(self):
+    # Returns a SyncLogSQL object, a saved instance from DB or
+    #   instantiated SQL object to be saved
+    # synclog_json_object should be a SyncLog instance
+    # if synclog_json_object._id
```
stray commented if statement

snopoke commented on corehq/ex-submodules/casexml/apps/phone/models.py:
```diff
@@ -354,6 +358,105 @@ def test_only_get_dependent_cases_on_phone(self):
+    domain = models.CharField(max_length=255, null=True, blank=True, default=None, db_index=True)
+    user_id = models.CharField(max_length=255, default=None, db_index=True)
+    date = models.DateTimeField(db_index=True, null=True, blank=True)
+    # needs to be a foreign key?
```
can delete this comment. I don't think it's worth making this a foreign key

snopoke commented on corehq/ex-submodules/casexml/apps/phone/models.py:
```diff
@@ -386,6 +489,12 @@ def last_for_user(cls, user_id):
+        save_synclog_to_sql(self)
+
+    def delete(self):
+        delete_synclog(self._id)
```
just noticing this now but can't you put these into the `AbstractSyncLog` class and then you won't need them here or in `SimplifiedSyncLog`

snopoke commented on corehq/ex-submodules/casexml/apps/phone/models.py:
```diff
@@ -1194,8 +1306,24 @@ def get_properly_wrapped_sync_log(doc_id):
+            return properly_wrap_sync_log(SyncLog.get_db().get(doc_id))
+        except ResourceNotFound:
+            pass
+        raise MissingSyncLog("A SyncLogSQL object with this synclog_id ({})is not found".format(
```
nit: rather have this wording generic "No sync log with..."

snopoke commented on corehq/warehouse/dbaccessors.py:
```diff
@@ -68,23 +68,14 @@ def get_synclog_ids_by_date(start_datetime, end_datetime):
+    return [
+        synclog_id.hex
+        for synclog_id in SyncLogSQL.objects.filter(
+            date__gt=start_datetime, date__lt=end_datetime
```
should be `date_lte` according to the docstring

snopoke reviewed (commented):
Final review revealed a few small comments but otherwise good to go. Let's address the comments today and then we can deploy after peak hours.

sravfeyn committed:
lint

sravfeyn committed:
remove not needed command

sravfeyn committed:
DRY

sravfeyn committed:
review

sravfeyn commented:
@snopoke I've made the changes requested.

sravfeyn committed:
merge remote

snopoke commented on corehq/ex-submodules/casexml/apps/phone/models.py:
```diff
@@ -842,9 +842,6 @@ def save(self, *args, **kwargs):
         save_synclog_to_sql(self)
 
-    def delete(self):
-        delete_synclog(self._id)
```
Sorry for the miscomms here. I meant to apply the same comment to `save` as well

sravfeyn replied:
The `save` is little bit different on both the models, no?

sravfeyn committed:
no special save for simplifiedsynclog

snopoke committed:
fix doc_type check

snopoke reviewed (approved):
Fixed tests. I think this is GTG. Thanks for pushing this through @sravfeyn

sravfeyn commented:
Thanks for the fix and for the patient and arduous review on this!

proteusvacuum commented:
@snopoke @sravfeyn FYI this broke prod, and it couldn't automatically be reverted. Please see https://github.com/dimagi/commcare-hq/pull/19562/files#r173953224
